### DISCUSSION TITLE
HW 1, Problem 2: Is water supposed to percolate upwards or no?

### DISCUSSION THREAD

**Discussion by shvnmttl** [williamgilpin/cphy#23] [homework]
This is a clarification question about the definition of percolation.

In the graph provided in the [hw1 jupyter notebook](https://github.com/williamgilpin/cphy/blob/main/hw/cellular_automata_complexity.ipynb) also attached here
![HWPercolationGraph](https://github.com/williamgilpin/cphy/assets/144622394/51ac1c23-4cab-49c9-a211-a7cc56a20b2a) 
the circled site could only be filled if water flowed upwards.
At the same time, the commented description of class PercolationSimulation says that,

> A simulation of a 2D directed percolation problem. Given a 2D lattice, blocked sites are denoted by 0s, and open sites are denoted by 1s. During a simulation, water is poured into the top of the grid, and allowed to percolate to the bottom. If water fills a lattice site, it is marked with a 2 in the grid. **Water only reaches a site if it reaches an open site directly above, or to the immediate left or right of an open site.**

This to me sounds contradictory to the above graph. Instead, taking the commented description of percolation to be correct, I suggest the correct graph should be,
![MyPercolationGraph](https://github.com/williamgilpin/cphy/assets/144622394/3752935f-d030-458c-ae74-bc1d4c058e4a)

Similarly, the given graph
![HWPercolationGraph2](https://github.com/williamgilpin/cphy/assets/144622394/72d8f0e7-18b5-4426-9a45-78b04f125830)
should be
![MyPercolationGraph2](https://github.com/williamgilpin/cphy/assets/144622394/22574dd4-18c0-48b2-a21d-be9c98bbaedb)

What do you think? Thanks!

**Comment by williamgilpin** [comment]
I very much agree. If you solve using DFS, it will automatically fill the caverns above the blue sites. That's how I generated my visualizations. But if you solve instead iteratively, but doing row 1, then row 2, etc, then you will get the other diagram.

Both methods will give the same $p_c$

To get the iterative method to draw a different picture, you can flip the lattice over and repeat the simulation, and then color sites blue if they got filled in either or both simulations. Or, you could modify the DFS solution to only go downwards (that would correspond to commenting out one line in the recursive call). The latter actually should give you a minor speedup in calculating $p_c$

**Reply by williamgilpin** [reply]
I'll clarify the description in next year's homework. The iterative solution is similar to BFS, and so I wanted to draw more attention to that anyway

**Reply by shvnmttl** [reply]
Ah got it! Thanks for the clarification! 
one thing that maybe I am not seeing too well is $p_c$ remaining the same. I would have thought that for smaller lattices (smaller $n$) the number of trajectories going down up and then back down that evaluate simulate to true could be significant. Because of which, it’s a bit counter intuitive for me that $p_c$ would be identical for both scenarios — DFS vs. iterative downward flow only.

**Reply by williamgilpin** [reply]
you're right, it also occurred to me that there's a way to construct a pathological case where the only way to percolate is by going up and then back down. So DFS will percolate, but the iterative won't. For example, if 0 denotes open sites and 1 denotes blocked sites, this will evaluate to True with DFS but not with the simple scan:

```
10111111
10110001
10000101
11111101
```

I'll think about changing the question and dropping the iterative solution for a proper BFS

**Reply by shvnmttl** [reply]
Thanks for the clarification :)

**Comment by shvnmttl** [comment]
Quick comment about a typo. The text above Todo for problem 2 says "fraction of open sites $p$." However, the description of the class in the code and the graphs given in the HW are made according to $p$ equal to the fraction of **closed** sites. I think the text above Todo for problem 2 is a typo.

Edit:
And, the axes for one of the graphs is flipped.
"Average site occupation probability" should be the $x$ axis and the "Standard deviation in site occupation probability" should be the $y$ axis.

**Reply by williamgilpin** [reply]
Oh dang good catch. Yes I will fix that, thanks for reading that carefully
